x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -115,96 +165,127 @@ private void loadModel(String modelId) {
 
     private void handleLoadSuccess(String modelId, TrainedModelConfig trainedModelConfig) {
         Queue<ActionListener<Model>> listeners;
-        Model loadedModel = new LocalModel(trainedModelConfig.getModelId(), trainedModelConfig.getDefinition());
+        LocalModel loadedModel = new LocalModel(trainedModelConfig.getModelId(), trainedModelConfig.getDefinition());
         synchronized (loadingListeners) {
             listeners = loadingListeners.remove(modelId);
             // If there is no loadingListener that means the loading was canceled and the listener was already notified as such
             // Consequently, we should not store the retrieved model
-            if (listeners != null) {
-                loadedModels.put(modelId, MaybeModel.of(loadedModel));
+            if (listeners == null) {
+                return;
             }
+            localModelCache.put(modelId, loadedModel);
```

Comment:
I think the implementation has a small race: if >1 threads call this with the same modelId/Config the model gets loaded twice, gets overwritten in the cache but also will be twice in memory as the loadModel instances are returned (line 180). If line 173 kicks in you also load the model superfluously.

I suggest to remove line 168 and use computeIfAbsent on line 176 instead of put. That ensures that the model is loaded only if necessary and exactly once.

Reply:
> If line 173 kicks in you also load the model superfluously.
That is true, but the loading already kicked off. This check is here for the scenario when a processor references a model_id, the loading starts, but the the processor is removed before it can be fully loaded. So, we should not cache it as nothing cares about it. 

If this method gets called more than once for the same modelId and loadedModel, it will not get added to cache twice. 

Since `loadingListeners` is a map by `modelId` even if more than one model load event has occurred (which I am not convinced is possible), the listeners are only notified once since the map entry is removed in line 170. 

The second thread to enter the `synchronized` block would see that `loadingListeners` does not have a key entry for `modelId` and will exit early without putting the model in cache.

Reply:
With twice in memory I did not mean the cache: this would be impossible. I mean the instances in memory. say A and B call this, both load it. Than A enters the `synchronized` block first, puts it into the cache and "returns" the model. Whoever that is uses instanceA. Than B enters the block, overwrites the cache entry and "returns" the model instanceB.
In the end you have instanceB in the cache and instanceA and instanceB in memory.

This whole thing is a bit constructed and probably not possible due to the call chain before that can happen.